itlwm/hal_iwn/ItlIwn.cpp

```diff
@@ -4040,42 +4077,93 @@ iwn_set_link_quality(struct iwn_softc *sc, struct ieee80211_node *ni)
 
-    i = 0;
-    if (ni->ni_flags & IEEE80211_NODE_HT) {
-        int txmcs;
-        for (txmcs = ni->ni_txmcs; txmcs >= 0; txmcs--) {
-            rinfo = &iwn_rates[iwn_mcs2ridx[txmcs]];
-            linkq.retry[i].plcp = rinfo->ht_plcp;
-            linkq.retry[i].rflags = rinfo->ht_flags;
-
-            if (ni->ni_htcaps & IEEE80211_HTCAP_SGI20)
-                linkq.retry[i].rflags |= IWN_RFLAG_SGI;
-
-            /* XXX set correct ant mask for MIMO rates here */
-            linkq.retry[i].rflags |= IWN_RFLAG_ANT(txant);
-
-            if (++i >= IWN_MAX_TX_RETRIES)
-                break;
-        }
-    } else {
-        int txrate;
-        for (txrate = ni->ni_txrate; txrate >= 0; txrate--) {
-            rinfo = &iwn_rates[wn->ridx[txrate]];
-            linkq.retry[i].plcp = rinfo->plcp;
-            linkq.retry[i].rflags = rinfo->flags;
-            linkq.retry[i].rflags |= IWN_RFLAG_ANT(txant);
-            if (++i >= IWN_MAX_TX_RETRIES)
-                break;
-        }
+    if (ic->ic_flags & IEEE80211_F_USEPROT)
+        if (sc->hw_type != IWN_HW_REV_TYPE_4965)
+            linkq.flags |= IWN_LINK_QUAL_FLAGS_SET_STA_TLC_RTS;
+    
+    if ((ni->ni_flags & IEEE80211_NODE_HT) &&
+        ieee80211_node_supports_ht_sgi20(ni)) {
+        ni->ni_flags |= IEEE80211_NODE_HT_SGI20;
+        sgi_ok = 1;
     }
+    
+    /*
+     * Fill the LQ rate selection table with legacy and/or HT rates
+     * in descending order, i.e. with the node's current TX rate first.
+     * In cases where throughput of an HT rate corresponds to a legacy
+     * rate it makes no sense to add both. We rely on the fact that
+     * iwm_rates is laid out such that equivalent HT/legacy rates share
+     * the same IWM_RATE_*_INDEX value. Also, rates not applicable to
+     * legacy/HT are assumed to be marked with an 'invalid' PLCP value.
+     */
+    j = 0;
+    ridx_min = iwn_rval2ridx(ieee80211_min_basic_rate(ic));
+    mimo = iwn_is_mimo_mcs(ni->ni_txmcs);
+    ridx_max = (mimo ? IWN_LAST_HT_RATE : IWN_LAST_HT_SISO_RATE);
+    for (ridx = ridx_max; ridx >= ridx_min; ridx--) {
+        uint8_t plcp = iwn_rates[ridx].plcp;
+        uint8_t ht_plcp = iwn_rates[ridx].ht_plcp;
+        
+        if (j >= IWN_MAX_TX_RETRIES)
+            break;
+        tab = 0;
+        rflags = 0;
+        if (ni->ni_flags & IEEE80211_NODE_HT) {
+            if (ht_plcp == IWN_RATE_HT_SISO_MCS_INV_PLCP)
+                continue;
+            /* Do not mix SISO and MIMO HT rates. */
+            if ((mimo && !iwn_is_mimo_ht_plcp(ht_plcp)) ||
+                (!mimo && iwn_is_mimo_ht_plcp(ht_plcp)))
+                continue;
+            for (i = ni->ni_txmcs; i >= 0; i--) {
+                if (isclr(ni->ni_rxmcs, i))
+                    continue;
+                if (ridx == iwn_mcs2ridx[i]) {
+                    tab = ht_plcp;
+                    rflags |= IWN_RFLAG_MCS;
+                    if (sgi_ok)
+                        rflags |= IWN_RFLAG_SGI;
```

Comment:
@pigworlds A weird thing is that the Tx performance drops seriously when SGI is enabled on 5GHz/40MHz channels, do you have the same experience? My card is N-6235.

Reply:
I didn't have the same experience with my Intel N-6205 connecting to Netgear R7800 router. It shows both Tx/Rx at 300.0 Mbit/s, 40 MHz, MCS 15, Short GI, while running iperf3 test.

I recall that with certain routers, the performance drops quickly when I move the laptop around, and stable when the laptop remains stationary. I don't remember specific SGI settings of those routers.

Reply:
I see, thank you.